enterprise/server/cmd/action_runner/main.go

```diff
@@ -0,0 +1,662 @@
+type actionRunner struct {
+	action        *workflowconf.Action
+	log           *invocationLog
+	bep           *buildEventPublisher
+	progressCount int32
+	username      string
+	hostname      string
+}
+
+func (ar *actionRunner) Run(ctx context.Context) error {
+	ar.log.Printf("Action:          %s", ar.action.Name)
+	ar.log.Printf("Triggered by:    %s to branch %q", *triggerEvent, *triggerBranch)
+	ar.log.Printf("Invocation ID:   %s", ar.bep.streamID.InvocationId)
+	ar.log.Printf("Invocation URL:  %s", invocationURL(ar.bep.streamID.InvocationId))
+
+	// NOTE: In this func we return immediately with an error of nil if event publishing fails,
+	// because that error is instead surfaced in the caller func when calling
+	// `buildEventPublisher.Wait()`
+
+	bep := ar.bep
+
+	startedEvent := &bespb.BuildEvent{
+		Id: &bespb.BuildEventId{Id: &bespb.BuildEventId_Started{Started: &bespb.BuildEventId_BuildStartedId{}}},
+		Children: []*bespb.BuildEventId{
+			{Id: &bespb.BuildEventId_Progress{Progress: &bespb.BuildEventId_ProgressId{OpaqueCount: 0}}},
+			{Id: &bespb.BuildEventId_WorkspaceStatus{WorkspaceStatus: &bespb.BuildEventId_WorkspaceStatusId{}}},
+			{Id: &bespb.BuildEventId_BuildFinished{BuildFinished: &bespb.BuildEventId_BuildFinishedId{}}},
+		},
+		Payload: &bespb.BuildEvent_Started{Started: &bespb.BuildStarted{
+			Uuid:            ar.bep.streamID.InvocationId,
+			StartTimeMillis: time.Now().UnixNano() / int64(time.Millisecond),
+		}},
+	}
+	if err := bep.Publish(startedEvent); err != nil {
+		return nil
+	}
+	if err := ar.flushProgress(); err != nil {
+		return nil
+	}
+	workspaceStatusEvent := &bespb.BuildEvent{
+		Id: &bespb.BuildEventId{Id: &bespb.BuildEventId_WorkspaceStatus{WorkspaceStatus: &bespb.BuildEventId_WorkspaceStatusId{}}},
+		Payload: &bespb.BuildEvent_WorkspaceStatus{WorkspaceStatus: &bespb.WorkspaceStatus{
+			Item: []*bespb.WorkspaceStatus_Item{
+				{Key: "BUILD_USER", Value: ar.username},
+				{Key: "BUILD_HOST", Value: ar.hostname},
+				{Key: "REPO_URL", Value: *repoURL},
+				{Key: "COMMIT_SHA", Value: *commitSHA},
+				{Key: "GIT_TREE_STATUS", Value: "Clean"},
+				// TODO: Populate GIT_BRANCH. Can't source this from the `trigger_branch` flag
+				// in the PR case, because that refers to the branch into which the PR would be
+				// merged, which doesn't reflect the currently checked out branch.
+			},
+		}},
+	}
+	if err := bep.Publish(workspaceStatusEvent); err != nil {
+		return nil
+	}
+
+	ps1End := "$"
+	if ar.username == "root" {
+		ps1End = "#"
+	}
+
+	stopFlushingProgress := ar.startBackgroundProgressFlush()
+	defer stopFlushingProgress()
+
+	for _, bazelCmd := range ar.action.BazelCommands {
+		args, err := bazelArgs(bazelCmd)
+		// Provide some info to help make it clear which output is coming
+		// from which bazel commands.
+		ar.log.Printf("\n%s%s@%s%s%s bazelisk %s", textCyan, ar.username, ar.hostname, textReset, ps1End, strings.Join(args, " "))
```

Comment:
hopefully, we can replace this with a multi-writer. But separately, I would remove the color-ization from the fmt strings and handle it in a different way. We import a library in splash.go that can simplify a lot of this, I think it does something like:

io.WriteString(writer, cyan("my cool text"))

Reply:
Done